staging/proj.finassistant/index.js

```diff
@@ -0,0 +1,365 @@
+// -*- mode: js; indent-tabs-mode: nil; js-basic-offset: 4 -*-
+//
+// Copyright 2019-2020 The Board of Trustees of the Leland Stanford Junior University
+//
+// Redistribution and use in source and binary forms, with or
+// without modification, are permitted provided that the following
+// conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+// 2. Redistributions in binary form must reproduce the above
+//    copyright notice, this list of conditions and the following
+//    disclaimer in the documentation and/or other materials
+//    provided with the distribution.
+// 3. Neither the name of the copyright holder nor the names of its
+//    contributors may be used to endorse or promote products derived
+//    from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+// FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+// COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+// INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+// (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+// OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Author: James zhuang <james.zhuang@cs.stanford.edu>
+"use strict";
+
+const Tp = require('thingpedia');
+const URL = "https://www.alphavantage.co/query?function=";
+const io = require('socket.io-client');
+const server = 'http://localhost:3000';
+const moment = require('moment');
+const fs = require('fs');
```

Comment:
Alright, so, right off the bat, I can tell you this architecture will not work. Almond skills can be executed in a variety of environments, most commonly in our server at almond.stanford.edu, and you won't have the ability to open ports or write to the local filesystem (well, you do have that ability, it's just that nothing else can see your local filesystem so it's not very useful).

Any data coming out of the skill needs to go through the Almond API, that is, it needs to be declared in the manifest and returned by the Thingpedia API.

Now, currently the Almond API doesn't expose the raw results unless you use the low-level ThingTalk execution API (which bypasses the natural language). But see https://github.com/stanford-oval/genie-toolkit/pull/420 for a PR that introduces exactly the API you need: you keep using the conversation API (stateful with websocket or stateless with REST and conversation ID), which takes text in and text out, but you also get a message every time Almond executes a ThingTalk program, with the result of that program. You can use that message as clue that the user wants to display a new chart, and display the chart.

You should move the logic for maintaining the dashboard (appending / replacing / clearing) to the server that the browser will connect to.

Reply:
Yes, will follow this approach and reorg a lot of the existing code